PerimeterXModule/Internals/Helpers/RemoteServer.cs

```diff
@@ -0,0 +1,194 @@
+			{
+				Stream clientStream = _context.Request.InputStream;
+				byte[] clientPostData = new byte[_context.Request.InputStream.Length];
+				clientStream.Read(clientPostData, 0, (int)_context.Request.InputStream.Length);
+
+				request.ContentType = _context.Request.ContentType;
+				request.ContentLength = clientPostData.Length;
+				Stream stream = request.GetRequestStream();
+				stream.Write(clientPostData, 0, clientPostData.Length);
+				stream.Close();
+			}
+
+			return request;
+
+		}
+
+		/// <summary>
+		/// Send the request to the remote server and return the response
+		/// </summary>
+		/// <param name="request">Request to send to the server </param>
+		/// <returns>Response received from the remote server
+		///           or null if page not found </returns>
+		public HttpWebResponse GetResponse(HttpWebRequest request)
+		{
+			HttpWebResponse response;
+
+			try
+			{
+				response = (HttpWebResponse)request.GetResponse();
+			}
+			catch (WebException e)
+			{
+				Debug.WriteLine("Failed to get response: " + e.Message, PxConstants.LOG_CATEGORY);
+				// Send 404 to client 
+				_context.Response.StatusCode = 404;
+				_context.Response.StatusDescription = "Page Not Found";
+				_context.Response.Write("Page not found");
+				return null;
+			}
+
+			return response;
+		}
+
+		/// <summary>
+		/// Return the response in bytes array format
+		/// </summary>
+		/// <param name="response">Response received
+		///             from the remote server </param>
+		/// <returns>Response in bytes </returns>
+		public byte[] GetResponseStreamBytes(HttpWebResponse response)
+		{
+			int bufferSize = 256;
+			byte[] buffer = new byte[bufferSize];
+			Stream responseStream;
+			MemoryStream memoryStream = new MemoryStream();
+			int remoteResponseCount;
+			byte[] responseData;
+
+			responseStream = response.GetResponseStream();
+			remoteResponseCount = responseStream.Read(buffer, 0, bufferSize);
+
+			while (remoteResponseCount > 0)
+			{
+				memoryStream.Write(buffer, 0, remoteResponseCount);
+				remoteResponseCount = responseStream.Read(buffer, 0, bufferSize);
+			}
+
+			responseData = memoryStream.ToArray();
+
+			memoryStream.Close();
+			responseStream.Close();
+
+			memoryStream.Dispose();
+			responseStream.Dispose();
+
+			return responseData;
+		}
+
+		/// <summary>
+		/// Set cookies received from remote server to response of navigator
+		/// </summary>
+		/// <param name="response">Response received
+		///                 from the remote server</param>
+		public void SetContextCookies(HttpWebResponse response)
+		{
+			_context.Response.Cookies.Clear();
+
+			foreach (Cookie receivedCookie in response.Cookies)
+			{
+				HttpCookie c = new HttpCookie(receivedCookie.Name,
+								   receivedCookie.Value);
+				c.Domain = _context.Request.Url.Host;
+				c.Expires = receivedCookie.Expires;
+				c.HttpOnly = receivedCookie.HttpOnly;
+				c.Path = receivedCookie.Path;
+				c.Secure = receivedCookie.Secure;
```

Comment:
Are we certain that by design these fields will _always_ be present on each HttpCookie (.Net sets default values)? Otherwise perhaps we need to verify that each such property is present (and not null) before assigning.

Reply:
Actually it was the first design but later on I understand that we dont need the set cookies as these are already sensitive headers so i removed the usage of SetContextCookies